src/txn/txn.c

```diff
@@ -1039,8 +1039,48 @@ __txn_fixup_prepared_update(
         hs_cursor->set_value(hs_cursor, &tw, tw.durable_stop_ts, tw.durable_start_ts,
           (uint64_t)WT_UPDATE_STANDARD, &hs_value);
         WT_ERR(hs_cursor->update(hs_cursor));
-    } else
-        WT_ERR(hs_cursor->remove(hs_cursor));
+    } else {
+        /*
+         * Remove the history store entry if checkpoint is not running. Otherwise, fix the history
+         * store entry with a prepared tombstone when the history store entry doesn't have its own
+         * stop timestamp.
+         */
+        if (txn_global->checkpoint_running) {
+            /* Don't update the history store entry if the entry already have a stop timestamp. */
+            if (fix_upd->type != WT_UPDATE_TOMBSTONE) {
+                /*
+                 * When a history store update start transaction is greater than the checkpoint
+                 * invisible transaction id, the durable timestamp of this update also must be
+                 * greater than the checkpoint timestamp, so there is no point in saving this
+                 * unstable update in the history store.
+                 */
+                if (fix_upd->txnid > txn_global->checkpoint_invisible_txn_id)
+                    WT_ERR(hs_cursor->remove(hs_cursor));
+                else {
+                    tw.durable_stop_ts = fix_upd->durable_ts;
+                    tw.stop_ts = fix_upd->start_ts;
+
+                    /*
+                     * Use checkpoint invisible transaction id, so that this tombstone is not
```

Comment:
"Set the stop transaction id of the time window to the checkpoint invisible transaction id. As such the tombstone won't be visible to rollback to stable, additionally checkpoint garbage collection cannot clean it up as it greater than the globally visible transaction id."

Reply:
Fixed.